File: web/app/src/pages/contacts/vcard.ts
import type { Contact, ContactInput } from "./types";

/** A contact parsed from a vCard file. Mirrors the editable ContactInput subset. */
export type ParsedContact = Omit<ContactInput, "starred"> & {
  starred?: boolean;
};

function emptyParsed(): ParsedContact {
  return {
    display_name: "",
    first_name: "",
    last_name: "",
    company: "",
    job_title: "",
    emails: [],
    phones: [],
    labels: [],
    notes: "",
  };
}

// vCard escaping per RFC 6350 §3.4: \n -> newline, and \\ \, \; literal.
function unescapeValue(v: string): string {
  return v.replace(/\\([\\,;nN])/g, (_, ch) =>
    ch === "n" || ch === "N" ? "\n" : ch,
  );
}

/**
 * Split a vCard content line into { name, params, value }.
 * Handles property params (e.g. EMAIL;TYPE=INTERNET:foo@bar) and grouping
 * prefixes (e.g. item1.EMAIL:...). The value is everything after the first
 * unescaped colon.
 */
function parseLine(
  line: string,
): { name: string; params: string[]; value: string } | null {
  const colon = line.indexOf(":");
  if (colon < 0) return null;
  const head = line.slice(0, colon);
  const value = line.slice(colon + 1);
  const parts = head.split(";");
  let name = parts[0] || "";
  // Strip a grouping prefix like "item1.EMAIL".
  const dot = name.lastIndexOf(".");
  if (dot >= 0) name = name.slice(dot + 1);
  return { name: name.trim().toUpperCase(), params: parts.slice(1), value };
}

function uniqPush(arr: string[], v: string) {
  const t = v.trim();
  if (t && !arr.some((x) => x.toLowerCase() === t.toLowerCase())) arr.push(t);
}

/**
 * Parse one or more vCards from raw .vcf text into ParsedContact records.
 * Supports BEGIN:VCARD/END:VCARD blocks with FN, N, EMAIL, TEL, ORG, TITLE,
 * NOTE, and CATEGORIES. Handles RFC line folding (continuation lines that
 * begin with a space or tab) and CRLF/LF line endings.
 */
export function parseVCards(text: string): ParsedContact[] {
  // Unfold folded lines (a CRLF followed by a space/tab continues the prior line).
  const unfolded = text.replace(/\r\n/g, "\n").replace(/\n[ \t]/g, "");
  const lines = unfolded.split("\n");

  const out: ParsedContact[] = [];
  let cur: ParsedContact | null = null;

  for (const raw of lines) {
    const line = raw.trimEnd();
    if (!line) continue;
    const upper = line.toUpperCase();
    if (upper === "BEGIN:VCARD") {
      cur = emptyParsed();
      continue;
    }
    if (upper === "END:VCARD") {
      if (cur) {
        if (!cur.display_name) {
          cur.display_name =
            `${cur.first_name} ${cur.last_name}`.trim() || cur.emails[0] || "";
        }
        out.push(cur);
      }
      cur = null;
      continue;
    }
    if (!cur) continue;
    const parsed = parseLine(line);
    if (!parsed) continue;
    const value = unescapeValue(parsed.value);
    switch (parsed.name) {
      case "FN":
        cur.display_name = value.trim();
        break;
      case "N": {
        // N: Family;Given;Additional;Prefix;Suffix
        const [family = "", given = ""] = value.split(";");
        if (given.trim()) cur.first_name = given.trim();
        if (family.trim()) cur.last_name = family.trim();
        break;
      }
      case "EMAIL":
        uniqPush(cur.emails, value);
        break;
      case "TEL":
        uniqPush(cur.phones, value);
        break;
      case "ORG":
        // ORG can be "Company;Department"; take the first component.
        cur.company = value.split(";")[0].trim();
        break;
      case "TITLE":
        cur.job_title = value.trim();
        break;
      case "NOTE":
        cur.notes = cur.notes ? `${cur.notes}\n${value}` : value;
        break;
      case "CATEGORIES":
        value.split(",").forEach((c) => uniqPush(cur!.labels, c));
        break;
      default:
        break;
    }
  }
  return out;
}

/** True if a parsed contact carries enough info to be worth creating. */
export function isMeaningful(c: ParsedContact): boolean {
  return Boolean(
    c.display_name ||
      c.first_name ||
      c.last_name ||
      c.emails.length ||
      c.phones.length,
  );
}

/** Build the create payload for a parsed vCard contact. */
export function toCreateInput(c: ParsedContact): Partial<ContactInput> {
  return {
    display_name: c.display_name,
    first_name: c.first_name,
    last_name: c.last_name,
    company: c.company,
    job_title: c.job_title,
    emails: c.emails,
    phones: c.phones,
    labels: c.labels,
    notes: c.notes,
    starred: c.starred ?? false,
  };
}

// ---- Duplicate detection & merge ----

function normName(c: Contact): string {
  return (c.display_name || `${c.first_name} ${c.last_name}`)
    .trim()
    .toLowerCase();
}

/**
 * A group of contacts considered duplicates of one another, keyed by the
 * signal that matched them (a shared email, or a shared display name).
 */
export interface DuplicateGroup {
  key: string;
  reason: "email" | "name";
  contacts: Contact[];
}

/**
 * Find groups of duplicate contacts. Two contacts are duplicates if they share
 * a case-insensitive email address, OR share a non-empty normalized display
 * name. Groups are merged transitively (A~B, B~C => {A,B,C}). Only groups with
 * 2+ members are returned.
 */
export function findDuplicates(contacts: Contact[]): DuplicateGroup[] {
  // Union-find over contact indices.
  const parent = contacts.map((_, i) => i);
  const find = (i: number): number => {
    while (parent[i] !== i) {
      parent[i] = parent[parent[i]];
      i = parent[i];
    }
    return i;
  };
  const union = (a: number, b: number) => {
    parent[find(a)] = find(b);
  };

  const byEmail = new Map<string, number>();
  const byName = new Map<string, number>();
  contacts.forEach((c, i) => {
    for (const e of c.emails) {
      const k = e.trim().toLowerCase();
      if (!k) continue;
      if (byEmail.has(k)) union(i, byEmail.get(k)!);
      else byEmail.set(k, i);
    }
    const n = normName(c);
    if (n) {
      if (byName.has(n)) union(i, byName.get(n)!);
      else byName.set(n, i);
    }
  });

  const groups = new Map<number, number[]>();
  contacts.forEach((_, i) => {
    const r = find(i);
    if (!groups.has(r)) groups.set(r, []);
    groups.get(r)!.push(i);
  });

  const result: DuplicateGroup[] = [];
  for (const idxs of groups.values()) {
    if (idxs.length < 2) continue;
    const cs = idxs.map((i) => contacts[i]);
    // A group is reported as "email" when any two members share an address;
    // otherwise the only thing tying them together is a shared display name.
    const sharesEmail = cs.some((c) =>
      c.emails.some((e) => {
        const k = e.trim().toLowerCase();
        return (
          k &&
          cs.some(
            (o) =>
              o !== c && o.emails.some((oe) => oe.trim().toLowerCase() === k),
          )
        );
      }),
    );
    result.push({
      key: cs[0].id,
      reason: sharesEmail ? "email" : "name",
      contacts: cs,
    });
  }
  return result;
}

/**
 * Merge a group of contacts into a single ContactInput: keeps the primary's
 * scalar fields (falling back to others when blank) and unions emails, phones,
 * and labels. The primary should be the contact you intend to keep/update.
 */
export function mergeContacts(
  primary: Contact,
  others: Contact[],
): ContactInput {
  const all = [primary, ...others];
  const emails: string[] = [];
  const phones: string[] = [];
  const labels: string[] = [];
  for (const c of all) {
    c.emails.forEach((e) => uniqPush(emails, e));
    c.phones.forEach((p) => uniqPush(phones, p));
    c.labels.forEach((l) => uniqPush(labels, l));
  }
  const firstNonEmpty = (pick: (c: Contact) => string): string => {
    for (const c of all) {
      const v = pick(c).trim();
      if (v) return v;
    }
    return "";
  };
  const notes = all
    .map((c) => c.notes.trim())
    .filter(Boolean)
    .join("\n\n");
  return {
    display_name:
      firstNonEmpty((c) => c.display_name) ||
      firstNonEmpty((c) => `${c.first_name} ${c.last_name}`),
    first_name: firstNonEmpty((c) => c.first_name),
    last_name: firstNonEmpty((c) => c.last_name),
    company: firstNonEmpty((c) => c.company),
    job_title: firstNonEmpty((c) => c.job_title),
    emails,
    phones,
    labels,
    notes,
    starred: all.some((c) => c.starred),
  };
}

File: internal/contacts/vcard.go
package contacts

import (
	"strings"
	"unicode"
)

// ParsedCard holds the fields extracted from a single vCard block.
type ParsedCard struct {
	DisplayName string
	FirstName   string
	LastName    string
	Company     string
	JobTitle    string
	Emails      []string
	Phones      []string
	Labels      []string
	Notes       string
}

// ParseVCards parses one or more vCard 3.0/4.0 records from raw .vcf text.
// It handles RFC 6350 line-folding (CRLF+SPACE continuation), multiple
// BEGIN:VCARD/END:VCARD blocks, FN, N, EMAIL, TEL, ORG, TITLE, NOTE, and
// CATEGORIES properties. Unknown properties are silently ignored.
func ParseVCards(text string) []ParsedCard {
	// Normalise CRLF → LF, then unfold continuation lines (a LF followed by
	// whitespace means the next line is a continuation of the previous one).
	text = strings.ReplaceAll(text, "\r\n", "\n")
	text = strings.ReplaceAll(text, "\r", "\n")
	text = strings.ReplaceAll(text, "\n ", "")
	text = strings.ReplaceAll(text, "\n\t", "")

	lines := strings.Split(text, "\n")
	var out []ParsedCard
	var cur *ParsedCard

	for _, raw := range lines {
		line := strings.TrimRightFunc(raw, unicode.IsSpace)
		if line == "" {
			continue
		}
		upper := strings.ToUpper(line)
		if upper == "BEGIN:VCARD" {
			cur = &ParsedCard{}
			continue
		}
		if upper == "END:VCARD" {
			if cur != nil {
				if cur.DisplayName == "" {
					cur.DisplayName = strings.TrimSpace(cur.FirstName + " " + cur.LastName)
					if cur.DisplayName == "" && len(cur.Emails) > 0 {
						cur.DisplayName = cur.Emails[0]
					}
				}
				out = append(out, *cur)
			}
			cur = nil
			continue
		}
		if cur == nil {
			continue
		}

		name, value := splitContentLine(line)
		if name == "" {
			continue
		}
		value = unescapeVCard(value)

		switch name {
		case "FN":
			cur.DisplayName = strings.TrimSpace(value)
		case "N":
			// N: Family;Given;Additional;Prefix;Suffix
			parts := strings.SplitN(value, ";", 5)
			if len(parts) >= 1 {
				cur.LastName = strings.TrimSpace(parts[0])
			}
			if len(parts) >= 2 {
				cur.FirstName = strings.TrimSpace(parts[1])
			}
		case "EMAIL":
			if v := strings.TrimSpace(value); v != "" {
				uniqAppend(&cur.Emails, v)
			}
		case "TEL":
			if v := strings.TrimSpace(value); v != "" {
				uniqAppend(&cur.Phones, v)
			}
		case "ORG":
			// ORG: Company;Department — take the first component.
			cur.Company = strings.TrimSpace(strings.SplitN(value, ";", 2)[0])
		case "TITLE":
			cur.JobTitle = strings.TrimSpace(value)
		case "NOTE":
			if cur.Notes != "" {
				cur.Notes += "\n" + value
			} else {
				cur.Notes = value
			}
		case "CATEGORIES":
			for _, cat := range strings.Split(value, ",") {
				if v := strings.TrimSpace(cat); v != "" {
					uniqAppend(&cur.Labels, v)
				}
			}
		}
	}
	return out
}

// SerializeVCards serialises a slice of Fields into vCard 3.0 text (CRLF
// line endings, one BEGIN:VCARD…END:VCARD block per contact).
func SerializeVCards(contacts []Fields, displayNames []string) string {
	var sb strings.Builder
	for i, f := range contacts {
		dn := f.DisplayName
		if dn == "" && i < len(displayNames) {
			dn = displayNames[i]
		}
		if dn == "" {
			dn = strings.TrimSpace(f.FirstName + " " + f.LastName)
		}
		if dn == "" && len(f.Emails) > 0 {
			dn = f.Emails[0]
		}
		sb.WriteString("BEGIN:VCARD\r\n")
		sb.WriteString("VERSION:3.0\r\n")
		sb.WriteString("FN:")
		sb.WriteString(escapeVCard(dn))
		sb.WriteString("\r\n")
		sb.WriteString("N:")
		sb.WriteString(escapeVCard(f.LastName))
		sb.WriteString(";")
		sb.WriteString(escapeVCard(f.FirstName))
		sb.WriteString(";;;\r\n")
		if f.Company != "" || f.JobTitle != "" {
			sb.WriteString("ORG:")
			sb.WriteString(escapeVCard(f.Company))
			sb.WriteString("\r\n")
			sb.WriteString("TITLE:")
			sb.WriteString(escapeVCard(f.JobTitle))
			sb.WriteString("\r\n")
		}
		for _, e := range f.Emails {
			sb.WriteString("EMAIL;TYPE=INTERNET:")
			sb.WriteString(escapeVCard(e))
			sb.WriteString("\r\n")
		}
		for _, p := range f.Phones {
			sb.WriteString("TEL:")
			sb.WriteString(escapeVCard(p))
			sb.WriteString("\r\n")
		}
		if len(f.Labels) > 0 {
			sb.WriteString("CATEGORIES:")
			for j, l := range f.Labels {
				if j > 0 {
					sb.WriteString(",")
				}
				sb.WriteString(escapeVCard(l))
			}
			sb.WriteString("\r\n")
		}
		if f.Notes != "" {
			sb.WriteString("NOTE:")
			sb.WriteString(escapeVCardNote(f.Notes))
			sb.WriteString("\r\n")
		}
		sb.WriteString("END:VCARD\r\n")
	}
	return sb.String()
}

// splitContentLine splits a vCard content line (possibly with property
// parameters) into a canonical property name and its value. The name is
// upper-cased and any grouping prefix (e.g. "item1.EMAIL") is stripped.
// Returns ("", "") if the line contains no colon.
func splitContentLine(line string) (name, value string) {
	colon := strings.IndexByte(line, ':')
	if colon < 0 {
		return "", ""
	}
	head := line[:colon]
	value = line[colon+1:]

	// Strip parameters: split on ';', the first part is the property name.
	parts := strings.SplitN(head, ";", 2)
	name = strings.ToUpper(strings.TrimSpace(parts[0]))

	// Strip a grouping prefix like "item1.EMAIL".
	if dot := strings.LastIndexByte(name, '.'); dot >= 0 {
		name = name[dot+1:]
	}
	return name, value
}

// unescapeVCard reverses vCard text escaping per RFC 6350 §3.4:
//
//	\n or \N → newline
//	\\ → backslash
//	\, → comma
//	\; → semicolon
func unescapeVCard(s string) string {
	if !strings.ContainsRune(s, '\\') {
		return s
	}
	var b strings.Builder
	b.Grow(len(s))
	i := 0
	for i < len(s) {
		if s[i] == '\\' && i+1 < len(s) {
			switch s[i+1] {
			case 'n', 'N':
				b.WriteByte('\n')
			case '\\':
				b.WriteByte('\\')
			case ',':
				b.WriteByte(',')
			case ';':
				b.WriteByte(';')
			default:
				b.WriteByte(s[i+1])
			}
			i += 2
			continue
		}
		b.WriteByte(s[i])
		i++
	}
	return b.String()
}

// escapeVCard escapes special characters in a vCard property value (general
// text fields: FN, ORG, TITLE, EMAIL, TEL, CATEGORIES items).
func escapeVCard(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, ",", "\\,")
	s = strings.ReplaceAll(s, ";", "\\;")
	return s
}

// escapeVCardNote escapes a NOTE value, additionally converting literal
// newlines to the vCard \n escape.
func escapeVCardNote(s string) string {
	s = escapeVCard(s)
	s = strings.ReplaceAll(s, "\n", "\\n")
	return s
}

// uniqAppend appends v to *slice only when it is not already present
// (case-insensitive comparison).
func uniqAppend(slice *[]string, v string) {
	lower := strings.ToLower(v)
	for _, existing := range *slice {
		if strings.ToLower(existing) == lower {
			return
		}
	}
	*slice = append(*slice, v)
}

// IsMeaningful returns true when a ParsedCard has enough data to be worth
// creating (has a name, email, or phone number).
func IsMeaningful(c ParsedCard) bool {
	return c.DisplayName != "" || c.FirstName != "" || c.LastName != "" ||
		len(c.Emails) > 0 || len(c.Phones) > 0
}
